src/client/SocketMost-Client.ts

```diff
@@ -0,0 +1,145 @@
+                    this.emit(Os8104Events.AllocResult, message)
+                    break
+                }
+                default:
+                    console.log("no match", event)
+            }
+        });
+    }
+
+    getPositions(): Promise<void> {
+        return new Promise((resolve, reject) => {
+            const data: MessageOnly = {
+                eventType: SocketTypes.GetNodePosition
+            }
+            this.sendAppMessage(data)
+            const positionTimeout = setTimeout(() => {
+                reject('Get Node Position Message Time Out')
+                console.log('position request timed out')
+            }, 20)
+            this.once('positionUpdate', () => {
+                console.log('resolving')
+                clearTimeout(positionTimeout)
+                resolve()
+            })
+        })
+    }
+
+    // parseMostMessage (message) {
+    //     let data = {
+    //         fBlockID: message.data.readUint8(0),
+    //         instanceID: message.data.readUint8(1),
+    //         fktID: (message.data.slice(2,4).readUint16BE() >> 4),
+    //         opType: ((message.data.readUint16BE(2) & 0xF)),
+    //         telId: (message.data.readUint8(4) & 0xF0) >>4,
+    //         telLen: (message.data.readUint8(4) & 0xF),
+    //         data: message.type > 0x01 ? message.data.slice(5, message.data.length - 1) : message.data.slice(5),
+    //         sourceAddrHigh: message.sourceAddrHigh,
+    //         sourceAddrLow: message.sourceAddrLow
+    //     }
+    //
+    //     let messageOut = {...data}
+    //     messageOut.data = [...message.data]
+    //     if(!this.networkMaster) {
+    //         if(message.fBlockID === 2) {
+    //             console.log("network master found")
+    //         }
+    //     }
+    //     return data
+    // }
+
+    sendAppMessage(data: SocketMostSendMessage | MessageOnly) {
+        this.client.write(JSON.stringify(data))
+    }
+
+    getMaster() {
+        this.client.write(JSON.stringify({eventType: 'getMaster'}) + '\r\n')
+    }
+
+    sendControlMessage(data: SocketMostSendMessage): Promise<void> {
+        return new Promise((resolve, reject) => {
+            this.client.write(JSON.stringify({eventType: 'sendControlMessage', ...data}))
+            const controlMessageTimeout = setTimeout(() => {
+                reject('sendMessage request timed out')
+            }, 50)
+            this.once('messageSent', () => {
+                console.log('resolving sent Message')
+                clearTimeout(controlMessageTimeout)
+                resolve()
+            })
+        })
+    }
+
+    allocate() {
+        this.client.write(JSON.stringify({eventType: 'allocate'}))
+    }
+
+    stream(data: Stream) {
+        this.client.write(JSON.stringify({eventType: 'stream', ...data}) + '\r\n')
+    }
+
+    retrieveAudio(data: RetrieveAudio) {
+        this.client.write(JSON.stringify({eventType: 'retrieveAudio', ...data}))
+    }
+}
```

Comment:
It seems quite hard to be type-safe when sending data over JSON strings. I think the UDP functionality is great if someone wants to use the driver in a different language besides javascript, however sending a json string between libraries kind of nullifies the purpose of TS?

Reply:
You could still have strict type checking up to the point where you call JSON.stringify(...)

You could have a method that receives a union type (https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) of all possible events you send (still has type checking to verify you are passing the right data) - and then calls stringify and sends - so you know that up to the point of calling stringify the type is correct

Reply:
Did this in the server, I guess I should do the same here, unsure why I didn't to be honest! Think I was exploring different ways to handle it.

https://github.com/rhysmorgan134/SocketMost/blob/085b641b2e038f598a7dce496c0295f5f740b141/src/server/SocketMost.ts#L168-L170